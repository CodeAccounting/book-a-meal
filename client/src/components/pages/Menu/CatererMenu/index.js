import { connect, } from 'react-redux';
import { bindActionCreators } from 'redux';
import CatererMenu from './CatererMenu';
import { fetchMenu, setCurrentDay } from '../../../../store/operations/menu';
import { logout } from '../../../../store/operations/auth';
import { toggleModal } from '../../../../store/actions/ui';

const mapStateToProps = state => ({
  isFetching: state.isFetching,
  meals: state.menu.meals,
  submitting: state.menu.working,
  submitError: state.menu.error,
  currentDay: state.menu.currentDay,
  metadata: state.menu.metadata
});

const mapDispatchToProps = dispatch => bindActionCreators({
  logout, fetchMenu, toggleModal, setCurrentDay
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(CatererMenu);
