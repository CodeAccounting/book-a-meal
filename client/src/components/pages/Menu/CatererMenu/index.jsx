import React, { Component, Fragment } from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import Modal from '../../../../containers/shared/Modal';
import MealCard from '../../../shared/MealCard';
import DatePicker from '../../../shared/DatePicker';
import View from '../../../shared/View';
import { userPropTypes, mealObjPropTypes } from '../../../../helpers/proptypes';
import CardGroup from '../../../shared/CardGroup';

/**
 * @exports
 * @class CatererMenu
 * @extends Component
 * @classdesc Creates CatererMenu Component
 * @returns {JSX} CatererMenu Component
 */
class CatererMenu extends Component {
  static propTypes = {
    ...userPropTypes,
    meals: PropTypes.arrayOf(mealObjPropTypes).isRequired,
    isFetching: PropTypes.bool.isRequired,
    logout: PropTypes.func.isRequired,
    fetchMenu: PropTypes.func.isRequired,
    currentDay: PropTypes.string.isRequired,
    submitError: PropTypes.oneOfType([PropTypes.string, PropTypes.shape({
      date: PropTypes.shape({
        value: PropTypes.string
      })
    })]),
    submitting: PropTypes.bool.isRequired,
    toggleModal: PropTypes.func.isRequired,
    setCurrentDay: PropTypes.func.isRequired,
  }

  static defaultProps = {
    submitError: null
  }

  /**
 * @memberof CatererMenu
 * @returns {JSX} CatererMenu Component
 */
  componentDidMount() {
    this.fetchMenu();
  }

  /**
   * @memberof CatererMenu
   * @param {string} date
   * @returns {JSX} CatererMenu Component
   */
  updateCurrentDate = async (date) => {
    await this.props.setCurrentDay(date);
    this.fetchMenu();
  }

  /**
   * @memberof CatererMenu
   * @param {object} metadata
   * @returns {func} fetchMenu
  */
  fetchMenu = (metadata = null) => {
    this.props.fetchMenu(this.props.currentDay, metadata);
  }

  /**
   * @memberof CatererMenu
   * @param {object} metadata
   * @returns {func} load more menu
  */
 loadMoreMenu = () => this.fetchMenu(this.props.metadata)

  /**
   * @memberof CatererMenu
   * @returns {JSX} CatererMenu Component
   */
  renderCatererMenu = () => {
    const now = moment().format('YYYY-MM-DD');
    const showMenuBtn = !moment(now).isAfter(moment(this.props.currentDay));

    const menu = this.props.meals.map(meal =>
      (<MealCard
        type="menu"
        key={meal.id}
        meal={meal}
      />));

    return (
      <Fragment>
        <div className="content-wrapper meals menu-meals">
          <div className="date d-none-md">
            <h2>{moment(this.props.currentDay).format('dddd[,] Do MMMM YYYY')} <span style={{ cursor: 'pointer' }}>&nbsp;&#9662;</span></h2>
            <DatePicker handleSelectDate={this.updateCurrentDate} />
          </div>
          <div className="top">
            {showMenuBtn && <button className="btn btn-pri" id="menu-modal-btn" onClick={() => this.props.toggleModal('menu')}>View Menu</button>}
          </div>
          <div className="page-heading">
            <h2>{'Meal Items on This Day\'s Menu'}</h2>
            <hr />
          </div>
          {this.props.meals.length === 0 && <p className="text-center">There are no Meal Items on this Menu</p>}
          {this.props.meals.length !== 0 &&
            <CardGroup
              items={menu}
              metadata={this.props.metadata}
              loadMore={this.loadMoreMenu}
            />}
        </div>
      </Fragment>
    );
  }


  /**
   * @memberof CatererMenu
   * @returns {JSX} CatererMenu Component
   */
  render() {
    const {
      user, meals, logout, submitting, submitError, isFetching
    } = this.props;

    return (
      <Fragment>
        <View
          showTime={false}
          user={user}
          logout={logout}
          type="menu"
          isFetching={isFetching}
          updateCurrentDate={this.updateCurrentDate}
        >
          {this.renderCatererMenu()}
        </View>
        <Modal
          meals={meals}
          submitting={submitting}
          submitError={submitError}
        />
      </Fragment>
    );
  }
}

export default CatererMenu;
