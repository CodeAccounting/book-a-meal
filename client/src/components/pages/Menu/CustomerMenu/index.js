import { connect, } from 'react-redux';
import { bindActionCreators } from 'redux';
import CustomerMenu from './CustomerMenu';
import { logout } from '../../../../store/operations/auth';
import { fetchMenu } from '../../../../store/operations/menu';

const mapStateToProps = state => ({
  isFetching: state.isFetching,
  meals: state.menu.meals,
  metadata: state.meals.metadata
});

const mapDispatchToProps = dispatch => bindActionCreators({
  logout, fetchMenu
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(CustomerMenu);
