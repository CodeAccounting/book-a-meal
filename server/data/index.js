import users from './users.json';
import meals from './meals.json';
import menu from './menu.json';
import orders from './orders.json';
import errors from './errors.json';

export default {
  users, meals, menu, orders, errors
};
