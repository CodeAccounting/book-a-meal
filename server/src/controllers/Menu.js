import uuidv4 from 'uuid/v4';
import moment from 'moment';
import Controller from './Controller';
import errors from '../helpers/errors.json';
import stringToArray from '../helpers/stringToArray';
import mealsDB from '../dummyData/meals';

/**
 * @exports
 * @class Menu
 * @extends Controller
 */
class Menu extends Controller {
  /**
   * Creates a new item
   * @method create
   * @memberof Controller
   * @param {object} req
   * @param {object} res
   * @param {object} data
   * @returns {(function|object)} Function next() or JSON object
   */
  create(req, res, data) {
    // generate random id
    data[`${this.type}Id`] = uuidv4();

    // convert meals array string to array
    data.meals = stringToArray(data.meals);

    this.database.push(data);

    // get full meals object from mealsDB
    const fullData = Object.assign(data);
    fullData.meals = Menu.getMealObject(fullData.meals);

    return res.status(201).send(fullData);
  }

  /**
   * Updates an existing item
   * @method update
   * @memberof Controller
   * @param {object} req
   * @param {object} res
   * @param {object} data
   * @returns {(function|object)} Function next() or JSON object
   */
  update(req, res, data) {
    const itemIndex = this.database
      .findIndex(item => item[`${this.type}Id`] === req.params[`${this.type}Id`]);

    // return 404 error if index isn't found ie meal option doesnt exist
    if (itemIndex === -1) {
      return res.status(404).send({ error: errors[404] });
    }

    if (this.database[itemIndex].date.toString() < moment().format().toString()) {
      return res.status(422).send({ error: 'Menu Expired' });
    }

    const oldItem = this.database[itemIndex];

    // convert meals array string to array
    data.meals = stringToArray(data.meals);

    // delete id and replace date from data
    delete data.menuId;

    data.date = oldItem.date;

    // update old meal with new meal and assign it to it's position in the array
    this.database[itemIndex] = Object.assign(oldItem, data);

    // get full meals object from mealsDB
    const fullData = Object.assign(this.database[itemIndex]);
    fullData.meals = Menu.getMealObject(fullData.meals);

    return res.status(200).send(fullData);
  }

  /**
   * Gets meals from mealsDB with ID
   * @method getMealObject
   * @memberof Controller
   * @param {meals} mealIDArray
   * @returns {array} Array of filled Menu
   */
  static getMealObject(mealIDArray) {
    return mealIDArray.map(mealID => mealsDB.find(mealItem => mealItem.mealId === mealID));
  }
}

export default Menu;
