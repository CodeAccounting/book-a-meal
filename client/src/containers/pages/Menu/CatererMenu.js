import { connect } from 'react-redux';
import CatererMenu from '../../../components/pages/Menu/CatererMenu';
import { fetchMenu, setCurrentDay } from '../../../actions/menu';
import { logout } from '../../../actions/auth';
import { toggleModal } from '../../../actions/ui';

const mapStateToProps = state => ({
  isFetching: state.isFetching,
  meals: state.menu.meals,
  submitting: state.menu.working,
  submitError: state.menu.error,
  currentDay: state.menu.currentDay,
  metadata: state.menu.metadata
});

export default connect(mapStateToProps, {
  logout, fetchMenu, toggleModal, setCurrentDay
})(CatererMenu);
