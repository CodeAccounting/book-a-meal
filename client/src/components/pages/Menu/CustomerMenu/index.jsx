import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import MealCard from '../../../shared/MealCard';
import Cart from '../../../shared/Cart';
import { userPropTypes, mealObjPropTypes } from '../../../../helpers/proptypes';
import View from '../../../shared/View';
import Notification from '../../../shared/Notification';
import checkShopOpen from '../../../../helpers/checkShopOpen';
import getOrderFromLocalStorage from '../../../../helpers/getOrderFromLocalStorage';
import updateLocalStorageOrder from '../../../../helpers/updateLocalStorageOrder';
import CardGroup from '../../../shared/CardGroup';
import './CustomerMenu.scss';

/**
 * @exports
 * @class CustomerMenu
 * @extends Component
 * @classdesc Creates CustomerMenu Component
 * @returns {JSX} CustomerMenu Component
 */
class CustomerMenu extends Component {
  static propTypes = {
    ...userPropTypes,
    meals: PropTypes.arrayOf(mealObjPropTypes).isRequired,
    isFetching: PropTypes.bool.isRequired,
    logout: PropTypes.func.isRequired,
    fetchMenu: PropTypes.func.isRequired,
  }

  /**
   * @constructor
   * @memberof CustomerMenu
   * @param {object} props
   * @returns {JSX} CustomerMenu Component
   */
  constructor(props) {
    super();

    this.state = {
      order: getOrderFromLocalStorage(props.user)
    };
  }

  /**
   * @memberof CustomerMenu
   * @returns {JSX} CustomerMenu Component
   */
  componentDidMount() {
    this.fetchMenu();
  }

  /**
   * @memberof CustomerMenu
   * @param {string} itemId
   * @returns {void}
   */
  getItemIndex = itemId => this.state.order.meals.findIndex(item => item.id === itemId)

  /**
   * @memberof CustomerMenu
   * @param {object} meal
   * @returns {func} addOrderItem|updateOrderItem
   */
  handleOrderMealClick = (meal) => {
    const itemIndex = this.getItemIndex(meal.id);

    if (itemIndex !== -1) {
      const newQuantity = this.state.order.meals[itemIndex].quantity + 1;

      const orderItem = this.changeOrderQuantity(meal.id, newQuantity);

      return this.updateOrderItem(orderItem);
    }

    return this.addOrderItem(meal);
  }

  /**
   * @memberof CustomerMenu
   * @param {object} meal
   * @returns {void}
   */
  addOrderItem = (meal) => {
    const item = {
      id: meal.id,
      title: meal.title,
      price: meal.price,
      quantity: 1
    };

    this.setState(prevState => ({
      order: {
        ...prevState.order,
        meals: [...prevState.order.meals, item]
      }
    }), () =>
      updateLocalStorageOrder(this.props.user.id, this.state.order));
  }

  /**
   * @memberof CustomerMenu
   * @param {object} orderItem
   * @returns {void}
   */
  updateOrderItem = (orderItem) => {
    const itemIndex = this.getItemIndex(orderItem.id);

    this.setState(prevState => ({
      order: {
        ...prevState.order,
        meals: [
          ...prevState.order.meals.slice(0, itemIndex),
          {
            ...prevState.order.meals[itemIndex],
            ...orderItem
          },
          ...prevState.order.meals.slice(itemIndex + 1)
        ]
      }
    }), () =>
      updateLocalStorageOrder(this.props.user.id, this.state.order));
  }

  /**
   * @memberof CustomerMenu
   * @param {string} itemId
   * @returns {void}
   */
  removeOrderItem = (itemId) => {
    this.setState(prevState => ({
      order: {
        ...prevState.order,
        meals: prevState.order.meals.filter(orderItem => orderItem.id !== itemId)
      }
    }), () =>
      updateLocalStorageOrder(this.props.user.id, this.state.order));
  }

  /**
   * @memberof CustomerMenu
   * @param {object} event
   * @param {string} itemId
   * @returns {func} updateOrderItem
   */
  handleQuantityInputChange = (event, itemId) => {
    const quantity = event.target.value <= 0 ? 1 : event.target.value;

    const orderItem = this.changeOrderQuantity(itemId, parseInt(quantity, 10));
    return this.updateOrderItem(orderItem);
  }

  /**
   * @memberof CustomerMenu
   * @param {string} itemId
   * @param {number} quantity
   * @returns {object} orderItem
   */
  changeOrderQuantity = (itemId, quantity) => {
    const orderItem = this.state.order.meals.find(item => item.id === itemId);

    orderItem.quantity = quantity;

    return orderItem;
  }

  /**
   * @memberof CustomerMenu
   * @param {object} metadata
   * @returns {func} fetchMenu
  */
  fetchMenu = metadata => this.props.fetchMenu(this.props.currentDay, metadata);

  /**
   * @memberof CustomerMenu
   * @param {object} metadata
   * @returns {func} load more menu
  */
  loadMoreMenu = () => this.fetchMenu(this.props.metadata)

  /**
   * @memberof CustomerMenu
   * @returns {JSX} CustomerMenu Component
  */
  renderMenu = () => {
    const { meals } = this.props;
    const menu = meals.map(meal =>
      (<MealCard
        type="customer"
        key={meal.id}
        meal={meal}
        orderMeal={() => this.handleOrderMealClick(meal)}
        inBasket={!!this.state.order.meals.find(item => item.id === meal.id)}
      />));

    return (
      <Fragment>
        {meals.length === 0 && <p className="text-center">{'There are no Meals on Today\'s Menu'}</p>}
        {meals.length !== 0 &&
          <CardGroup
            items={menu}
            metadata={this.props.metadata}
            loadMore={this.loadMoreMenu}
          />}
      </Fragment>
    );
  }

  /**
   * @memberof CustomerMenu
   * @returns {JSX} CustomerMenu Component
  */
  render() {
    const { user, logout, isFetching } = this.props;
    const isShopOpen = checkShopOpen();

    return (
      <View user={user} logout={logout} type="menu" isFetching={isFetching}>
        <div className="meals user-meals">
          <div className="user-menu">
            <div className="main-menu">
              <div className="page-heading">
                <h2>{'Today\'s Menu'}</h2>
                <hr />
              </div>
              {!isShopOpen && <Notification message="Ordering is only available between 8:30am and 4:00pm. Please check back later." />}
              {this.renderMenu()}
            </div>
            <Cart
              order={this.state.order.meals}
              handleQuantityInputChange={this.handleQuantityInputChange}
              removeOrderItem={this.removeOrderItem}
            />
          </div>
        </div>
      </View>
    );
  }
}

export default CustomerMenu;
