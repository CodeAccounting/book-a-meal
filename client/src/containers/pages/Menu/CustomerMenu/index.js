import { connect, } from 'react-redux';
import { bindActionCreators } from 'redux';
import CustomerMenu from '../../../../components/pages/Menu/CustomerMenu';
import { logout } from '../../../../actions/auth';
import { fetchMenu } from '../../../../actions/menu';

const mapDispatchToProps = dispatch => bindActionCreators({
  logout, fetchMenu
}, dispatch);

export default connect(null, mapDispatchToProps)(CustomerMenu);
