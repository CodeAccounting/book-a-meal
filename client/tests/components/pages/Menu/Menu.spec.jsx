import React from 'react';
import Menu from '../../../../src/components/pages/Menu';
import { caterer, customer } from '../../../setup/data';

const { now } = Date;

describe('Menu', () => {
  beforeAll(() => {
    Date.now = jest.fn(() => new Date().setMilliseconds(0));
  });

  afterAll(() => {
    Date.now = now;
  });

  it('renders correctly: Caterer Menu', () => {
    const shallowWrapper = shallow(<Menu type="caterer" user={caterer} />);

    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('renders correctly: Customer Menu', () => {
    const shallowWrapper = shallow(<Menu type="customer" user={customer} />);

    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });
});
