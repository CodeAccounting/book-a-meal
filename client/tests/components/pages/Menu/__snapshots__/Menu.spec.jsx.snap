// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correctly: Caterer Menu 1`] = `
<React.Fragment>
  <Connect(CatererMenu)
    type="caterer"
    user={
      Object {
        "address": "$, Herbert Macaulay Way, Yaba, Lagos",
        "businessName": "Food Circle",
        "email": "food@circle.com",
        "id": "a09a5570-a3b2-4e21-94c3-5cf483dbd1ac",
        "phoneNo": "08166557788",
        "role": "caterer",
      }
    }
  />
</React.Fragment>
`;

exports[`Menu renders correctly: Customer Menu 1`] = `
<React.Fragment>
  <Connect(CustomerMenu)
    type="customer"
    user={
      Object {
        "email": "emiola@olasanmi.com",
        "firstname": "Emiola",
        "id": "a09a5570-a3b2-4e21-94c3-5cf483dbd1ac",
        "lastname": "Olasanmi",
        "role": "customer",
      }
    }
  />
</React.Fragment>
`;
